### PR TITLE
[TS] Add generic type with default to Group children

### PR DESCRIPTION
**Pull request by xawill**
Based on a [discussion on Twitter](https://twitter.com/squidfunk/status/1273994680955518981), I suggest those modifications in the TypeScript definitions to introduce generics in Group children. This is a follow-up work of #19299 which introduced generics for Geometry and Material in Geometry types.

I also changed the type of the `type` property since people will be more likely to override Group and therefore have their own `type` string instead of 'Group'. I don't know if this is a good idea, but I think it would reduce the number of adaptations in an existing codebase to adopt this TS "feature". It is the case for one of my codebases at least.

Let me know what you think, especially @donmccurdy, since you suggested me the PR. Thanks !

### CODE REVIEW

**Comment by squidfunk** [comment]
I'd recommend defining it like:

``` ts
export class Group<TChildren extends Object3D[] = Object3D[]>  extends Object3D {
  children: TChildren
}
```

This would allow for the definition of element collections:

``` ts
const group: Group<SomeObject[]>
group.children[0] // => SomeObject
group.children[1] // => SomeObject
...
group.children[n] // => SomeObject
```

...as well as tuple types (i.e. ordered arrays) 😊

``` ts
const  group: Group<[SomeObject, SomeOtherObject]>
group.children[0] // => SomeObject
group.children[1] // => SomeOtherObject
group.children[2] // => never
```

**Comment by xawill** [comment]
Thank you for the suggestion ; it makes sense to me. I definitely get the added value in the case of tuple types, but I think my proposal is the same as yours @squidfunk in the case of element collections, right ?

**Comment by xawill** [comment]
Actually while starting to implement your proposal I am having mixed feelings about it (or maybe I am just showing my limitations in TypeScript): what about the `object` parameter of the `attach` method, for instance, with your suggestion ? It would be of type `Object3D` and we would therefore loose the type information of your children.

It looks to me that this is either we "force" the Group to have only children of the same type and we can type those methods with the actual child type or we allow for heterogeneous children but those methods will have an unrestricted type `Object3D`.

After this little more thinking, I am more in favour of my suggestion as I think it suits more use cases (included the one you outlined in your original tweet). But I have no strings attached to it, so I guess @donmccurdy 's output on this will be decisive :)

**Comment by squidfunk** [comment]
@xawill for element collections it's exactly the same, with the upside that it supports tuple types. `TChild`, the collection element type, can be extracted with a conditional type:

``` ts
type TupleType = [Mesh, LineSegments]
type TupleValueType<T> = T extends Array<infer U> ? U : never

TupleValueType<TupleType> // => Mesh | LineSegments
```

**Comment by xawill** [comment]
I get it now. Thank you very much @squidfunk for educating me today :)

I updated the PR with your suggestion. Let me know if you still see any remaining potential refinements.

**Comment by squidfunk** [comment]
There's a lot of duplication. You could either move the conditional type definition outside of the group, or:

``` ts
export class Group<
  TChildren extends Object3D[] = Object3D[],
  TChild = TChildren extends Array<infer U> ? U : never
>  extends Object3D {
  children: TChildren
  ...
  attach( object: TChild ): ...
  ....
}
```

However, thinking about it, with tuples, the methods that allow to add or remove objects make no real sense, as it shouldn't be allowed because tuples are more or less static. For this reason, I'm considering if we should maybe just drop the whole tuple support and I continue to use some hacks (interfaces or classes) to support tuples in the groups I'm using 😊

For example, I also found out that when extending from `Group`, I could just use:

``` ts
export class SpectrumContainer extends Group {
  public declare children: [
    SpectrumLineGroup,
    SpectrumPartitionGroup
  ]
}
```

The `declare` is necessary so that TypeScript doesn't complain that `children` is not explicitly set in the constructor, as we have to use `add`.

**Comment by xawill** [comment]
I agree that the duplication is not very elegant. But from what I know, both your suggestions imply introducing a second generic type, right ? Either at the Group declaration (outlined in your code snippet) or at the new conditional type declaration outside of `Group` (as I am pretty sure this is not possible to define types inside classes, unfortunately). I don't find it very elegant neither.

I am not sure we should drop the heterogeneous collection support. Indeed, the tuple use case you mention is not the best one, but we can still have heterogeneous arrays of `Mesh` and `LineSegments`, for instance, or `SomeType extends Mesh` and `SomeOtherType extends Mesh`. So I think this is worth it trying to support this. But right now I don't see a more elegant solution than the current one, do you ?

**Comment by xawill** [comment]
Do you have any additional comments to make to this @squidfunk ? If not, I'll propose the modifications as it is now, hoping them to be included in the next release.

**Comment by squidfunk** [comment]
@xawill LGTM. We can refine it later on if necessary.

**Comment by xawill** [comment]
Thanks a lot for your help!

**Comment by xawill** [comment]
@donmccurdy This PR is ready for review. (Sorry if you already received a notification for this. I don't want to push you; I am just trying to have this merged in r123.)

**Comment by donmccurdy** [comment]
Sorry, this is more advanced use of TypeScript than I know how to review. I'm not familiar with using ternaries and `infer` in type definitions. If someone can test this in their TypeScript project (with `npm link` or similar) and confirm it is working as expected and backward compatible, then it is fine with me.

**Comment by mrdoob** [comment]
I have not been able to read the reasoning for this but... isn't it strange that we're re-defining methods that `Group` inherits from `Object3D`?

**Comment by xawill** [comment]
> Sorry, this is more advanced use of TypeScript than I know how to review. I'm not familiar with using ternaries and `infer` in type definitions. If someone can test this in their TypeScript project (with `npm link` or similar) and confirm it is working as expected and backward compatible, then it is fine with me.

@donmccurdy Thank you for your reply. I just did it on my commercial medium-size project using a lot of three.js and it works fine both were I use the generics proposed in this PR and where I don't. But I guess you would need a proof of that and I don't really see how I would do it... Does three.js have a test suite for TypeScript ?

> I have not been able to read the reasoning for this but... isn't it strange that we're re-defining methods that `Group` inherits from `Object3D`?

@mrdoob Well, I can do the same on the `Object3D` class if you prefer. I understood from the docs that the `Group` class was the one you strongly suggest us to use when we want to add children to objects. At least, this is the one I use in my codebase. I guess the answer to your question is the same as the answer to the question "Why does the `Group` class exists if the `Object3D` one does exactly the same ?" And I don't have the answer to this question. Retro-compatibility, maybe ? If yes, then indeed, as this is a new feature, I don't have to do it in the `Group` class. Your call; I am fine with both. The only thing is that personally I would have to change the declaration of all my current groups to `Object3D` prior to using the added value of this PR, which I find less declarative code-wise.

**Comment by donmccurdy** [comment]
> But I guess you would need a proof of that and I don't really see how I would do it... Does three.js have a test suite for TypeScript?

I'm pretty willing to rely on TS users to tell us what works best here, similarly to if the typings were external in DefinitelyTyped. Unfortunately there doesn't seem to be any really robust way to validate `d.ts` files against the source JavaScript, which is frustrating. We do have `npm run test-lint` and `npm run lint-examples` which do some basic tsc checks against the typings if i understand correctly.

**Comment by xawill** [comment]
Do you mean that you are waiting for additional feedback from other TS users or that you are taking my word for it ?

Or maybe you want to wait to find a formal way to validate the changes ?

I understand from #19940 that TS validation against the source JS is indeed a sensitive topic. I just want to know if/how I can help to move forward.

**Comment by donmccurdy** [comment]
> Or maybe you want to wait to find a formal way to validate the changes ?

I don't think it's possible (without converting the JS to TS or Flow or Closure types, which we don't want to do) so there's no use to wait on it unfortunately. If you and @squidfunk are comfortable with this PR then no objection from me.

**Comment by mrdoob** [comment]
> Why does the `Group` class exists if the `Object3D` one does exactly the same ?

It helps `WebGLRenderer` to pretty much ignore the object:

https://github.com/mrdoob/three.js/blob/0d83541c6bba944636fb6350b9b57a045060b997/src/renderers/WebGLRenderer.js#L1125-L1129

We can't do that with `object.isObject3D`.

**Comment by xawill** [comment]
Thank you both for your explanations.
I think @squidfunk already gave [his validation](https://github.com/mrdoob/three.js/pull/19702#issuecomment-729681969) earlier, so I rebased the PR and you can merge whenever you want.

**Comment by mrdoob** [comment]
I still don't understand why this needs to be added to `Group`.
It looks to me that adding these to `Object3D` would be more appropriate.

**Comment by xawill** [comment]
There are two reasons.

The first one is the one I mentioned above: I thought we should prefer to use `Group` for grouping objects and not `Object3D`. This is written in the `Object3D` docs :

> Note that this can be used for grouping objects via the .add( object ) method which adds the object as a child, however it is better to use Group for this.

Again, I understand that the methods `add`, `remove`, etc... must remain in `Object3D` for backward-compatibility, but I think they should eventually exclusively belong in `Group`.

The second reason is accessory, as it is related to a TypeScript issue and I hate having to change anything in a software architecture just because TypeScript doesn't let us do what we want, but I can still mention it: circular default. The issue is that if I do the same thing as I suggest in this PR but for `Object3D`, we would have `Object3D<TChildren extends Object3D[] = Object3D[]> extends EventDispatcher` which is not allowed. Indeed, the default type in the generic is `Object3D[]` which is the class itself... We could break the circle with `Object3D<any>[]` as default type, but this would not be correct as far as Three.js is concerned, as we cannot have for instance an object of type `HTMLElement` (which doesn't extends `Object3D`) as a child of `Object3D`.

To recap, I don't currently know how to do it on `Object3D`, but this is not the reason why I did it on `Group`. I am convinced (but again, I, as a very minor contributor, have nothing to say about the architecture of your library so this is up to you) that all those methods to manually group children belong in `Group` and not in `Object3D`. Personally, I don't see why this would be inelegant to do it in `Group`. Overriding methods is common practice in all OOP language I know.

Let me know what you think. I am always open to discussion, especially since I want to find a solution to have this merged, as it would add a lot of syntactic sugar (of the same kind as when generics was introduced for `Geometry | BufferGeometry` and `Material | Material[]`) to the main project I use three.js for.

**Comment by xawill** [comment]
The more I try to find a solution to have it working inside `Object3D`, the more I think we should keep it in `Group`. I just tried a possible solution to do it but I stumbled upon another issue: the need to redefine the generic type on all the subclasses of `Object3D`. Indeed, if we implement my suggestion in `Object3D`, it means that we should be able to define the type of children of all subclasses of `Object3D` (the point of inheritance). But just inheriting from `Object3D` would not provide the generic type on the subclass. We would have to define what I did on `Group` here for all subclasses of `Object3D`. This leads to a lot of maintainability issues and I don't think we should do down this path.

So to clarify my use case, I want to be able to group multiple (subclasses of) `Object3D` in a `Group` and having the TypeScript type inference when I iterate over the children of the `Group`, for instance. Currently, I have to cast the type of `Group.children`, as it is `Object3D[]`.

I understand it might be more ideal to do it on `Object3D`, but this is currently not possible with the current version of Typescript and/or not maintenable, to my opinion. So I guess we could take my PR as a partial implementation of this ideal goal which I would be pleased to do once possible in TypeScript in a more elegant way.

**Comment by mrdoob** [comment]
But if this was inside `Object3D` this would automatically work for `Mesh`, `Line`, `Points`, ... No?

**Comment by xawill** [comment]
No, without addition in the subclass (say `Group`, for instance), it would take the default generic type of the superclass (`Object3D`), which is `Object3D` here. So this would compile, but not add the benefits of my suggestion for the subclasses, even if you especially write the generic type in the instantiation of the subclass (`new Group<AnyCustomSubclassOfObject3D>()`, for instance). Actually, this last piece of code would not compile with my suggestion only inside `Object3D` and nothing changed anywhere else (I just tested it, the error message is "Expected 0 type arguments, but got 1", regarding the generic type `AnyCustomSubclassOfObject3D `).

**Comment by mrdoob** [comment]
I don't know... This doesn't feel like the right solution. And I lack the time to learn TypeScript types at the moment.

**Comment by xawill** [comment]
Fair enough, but you get my use case, right ?

**Comment by xawill** [comment]
Trying harder to have this PR merged, I found a workaround which surprisingly works to break the recursion mentioned above:

```
type AliasObject3D = Object3D;
export class Object3D<TChild extends Object3D = AliasObject3D> extends EventDispatcher { /* Implementation */ }
```

I tested it, it works for me and satisfies my use case. I am then willing to update my PR to have this inside `Object3D`.

Remains the caveat regarding the fact that we have to edit all subclasses of `Object3D` where we want to implement this addition (at least `Group`, as far as I am concerned) with the following:

```
export class GroupOrAnySubclassOfObject3D<T extends Object3D = Object3D> extends Object3D<T> { /* Implementation */ }
```

How do you want me to do it @mrdoob ? Do I do the change on all subclasses of `Object3D` (including for instance `Bone`, `DirectionalLightHelper`, etc) ? Or only a subset of subclasses (if yes which ones) ? Or is this PR still a no-go for you as described here ?

**Comment by mrdoob** [comment]
For now would be good to see how this looks like in `Object3D` and then the changes needed for `Mesh`.

**Comment by xawill** [comment]
Great! Here it is: the first commit is the original one on `Group`, the second is the same on `Object3D` only and the third commit is what it would look like on any subclass of `Object3D` we would like to have this addition with the example on `Group` (the most useful to me) and `Mesh` (the one you requested). 

Personally, I think that if we do it on `Mesh`, we should at least do it for `Points` and `Line`, for coherence. All other subclass in the (big) hierarchy of the library (like `MMDPhysicsHelper`, for instance) look way less useful. 

Let me know how you want to move on from this and if you want me to squash the commits for this PR.

**Comment by mrdoob** [comment]
This feels better to me.

@squidfunk @donmccurdy what do you guys think?

**Comment by squidfunk** [comment]
Do tuple styles still work? As denoted in https://github.com/mrdoob/three.js/pull/19702#issuecomment-647322168, I find tuple types to be very useful when using groups.

**Comment by xawill** [comment]
@squidfunk I can add them back if really needed. But you suggested afterwards in [#19702 (comment)](https://github.com/mrdoob/three.js/pull/19702#issuecomment-647579265) to remove "the whole tuple support" as the `add`, `remove` and `attach` methods would not be very meaningful for tuples.

I follow you 100% on this. I prefer to match the current three.js API and potentially find a solution in a later stage to have a tuple support with the "conditional removal" of those methods in the case a tuple is passed as `children`.

As discussed before, I also removed the tuple support as this was either introducing a second useless generic type or being very verbose. I think the PR as it is currently has a higher probability of being merged someday without the tuple support (for now).

**Comment by squidfunk** [comment]
@xawill my bad 🤦‍♂️ I haven't revisited this topic for a long time so I only skimmed through the comments. Thus, LGTM!

**Comment by DavidPeicho** [comment]
I am not sure to understand what is the benefit of this PR?

I think such a change makes sense **only** if you are building your own `Object3D` hierarchy, and then `TChild` is basically your parent class. However, the way Three.js is built anyway you can't really do it.

I love having stronger types (and so better IntelliSense), but I don't see any value here.

**Comment by mrdoob** [comment]
I opted for asking on Twitter for opinions about this issue:
https://twitter.com/threejs_org/status/1354019157923266561

**Comment by xawill** [comment]
@DavidPeicho Well, `TChild` is not necessarily your parent class. But this is not the point. The point is to have it typed correctly without having to cast all the time `Object3D.children[i]` to `TChild` when you want to get a child of your `Object3D`.

Let me explain again my use case more concretely. Let's say I want to display a house in a scene. I have an `Object3D` for the house which is imported from a three.js loader. I then "parse" this model, which is provided by the user, to gather some information about the geometry and add some 3D objects on top of them. I add for instance the roof framework below the tiles of the roof which is an `Object3D` with some children, etc. Let's say your house has two roof planes (but it is often more), I group the roof framework of the two houses as a `Group` of `RoofFramework` which I add to my house `Object3D` hierarchy. My `RoofFramework` class extends `Object3D` but also adds a lot of capabilities (i.e. custom methods and properties) to the roof framework.

Later one, let's say I need to iterate over all the roof frameworks of my house (i.e. a `Group`). As said above, I'll have to iterate over `roofFrameworks.children[i]` which is of type `Object3D`. But I need a `RoofFramework` type to use directly the additional capabilities mentioned above.

And I use this "pattern" a lot to build my hierarchy. This means I have a lot of casting from `Object3D` to `RoofFramework` or any other specific type I defined. This would be so much easier if I could directly define the group `roofFrameworks` as `Group<RoofFramework>`. I would have directly the type `RoofFramework` when I do `roofFrameworks.children[i]`.

The rationale behind this PR is **exactly** the same as the one that introduced generics for `BufferGeometry | Geometry` and `Material | Material[]` in `Mesh`.

I hope this is clearer. Let me know if not.

**Comment by DavidPeicho** [comment]
I understand the issue and that comes back to what I said in a way.

The thing is: that's not an issue. Semantically, what Three.js is doing is right. an `Object3D` has a list of children that are implementing the `Object3D` protocol. That doesn't mean every object has the same type. `Group` hasn't been created to host a single type and that's causing your issue.

I don't think modifying Three.js like that makes sense. What @squidfunk does is the way to go on **your** side, i.e., in your app or framework.

By doing that:

```ts
export class Object3D<TChild extends Object3D = AliasObject3D> extends EventDispatcher {
    children: TChild[];
}
```

You are specifiying that your object will host a single type **as well as its sub types**. What happens now if you create your own
type `class MyType extends Object3D` that you want to add to a parent? What happens is that I can still add any other Object3D that will not be a descendant of `TChild`.

The way to go is to re-type your own `Object3D` in your own project using interfaces.

**Comment by xawill** [comment]
Whatever, I created this PR as a suggestion. I spent way too much time on it from the benefits it might give. This is a frustrating experience but I understand your point. Anyway, either merge or close.

**Comment by DavidPeicho** [comment]
Sorry if I sounded harsh, that wasn't my intention. My point is just that unfortunately, this can't really be solved. There are just now good ways of doing it by the nature of the hierarchy and the fact that `children` in the end should accept a wide range of objects.

I think what's been proposed on Twitter is really clean:

```ts
export interface Partition extends Mesh {
}
```

As far as I know,  there will be no generic enough way to have that in the library that would work for all cases (imagine for instance someone that as a `children` array of two `Object3D` different type.

**Comment by xawill** [comment]
Thank you for your clarification. I definitely grasp the complexity of having **the** perfect solution for that. I encountered the difficulty of implementing those recursive types in TS as you might have understood from the discussion above. I personally think that sometime there is no perfect solution and having one that satisfies say 90% of TS users which where using hacks on their side until now is better than status quo, even if 10% of users with very specific needs would have to **continue to** hack a little to suit their use case (providing 100% backwards compatibility).

I am no TS expert, but I think this implementation is 100% backwards compatible. Until now, there were no generic types, so no one is using them and the default type I used for the generics, `Object3D` is the most generic one accepted by the three.js library. 

I understand your suggestion on how I should solve the issue on my side. I am just not convinced that creating a huge number of interfaces (one per group, in my case) is cleaner than casting where needed. Unlike generics directly inside the library.

With the constraint that `TChild` **must** extend `Object3D`, you can do everything you currently can.

**Comment by DavidPeicho** [comment]
I don't think most of the people want to have an object3D whose children are bound to a finer Object3D hierarchy, that sounds more like the 10% of cases.

It also doesn't fix your issue that I can add anything to `children` that may not match your `TChild`, i.e., that is only conform with `Object3D`.

In your case, what you want would be something more like:

```ts
class MyObject3D extends Object3D {
} 
```

And then Three.js would need to handle Object3D like:

```ts
class ConstrainedObject3D<Base = Object3D, Child extends Base> {

  children: Child[]

}
```

But that will become quickly umanageable.

**Comment by xawill** [comment]
> I don't think most of the people want to have an object3D whose children are bound to a finer Object3D hierarchy, that sounds more like the 10% of cases.

You are right, that is why the generic type is optional (defaulted to an `Object3D` hierarchy, which is the current state of the library).

The 90% of people I mention are the one willing to use the generics for strongly type children of `Object3D` without a more complexe sub-hierarchy. This was related to your comment:
> What happens now if you create your own type class MyType extends Object3D that you want to add to a parent?

**Comment by xawill** [comment]
But again, you are the TS expert and @mrdoob values your feedback. I think we have different points of view and I don't think this is worth trying to convince you.

I understand @mrdoob wants the best for three.js, including regarding TS (and thank you for having merged the TS declarations in the repo), and this is why three.js is so awesome. But in my opinion the debate is not worth such a minor addition. I mean, I am not trying to remove `Geometry` from the library! 😆 So, again, we both made our points, I think @mrdoob has enough information to take an informed decision regarding merging or closing this PR.

**Comment by DavidPeicho** [comment]
Yes, my issue is more that it's not really "correct", the addition itself is a few lines yes. It works for what you are doing but doesn't make sense otherwise (maybe I am wrong). Same for the `Mesh` and `material` properties actually, but this is already merged :D

**Comment by xawill** [comment]
Thank you for acknowledging that my PR is very similar to the addition of generics for `geometry` and `material`. Because with that PR merged, this doesn't sound coherent to tell me that mine is not merged because such generics are not the responsibility of three.js to care about...

**Comment by mrdoob** [comment]
Thanks for the conversation guys 🙏

The main issue is that no developer has offered to take ownership of the d.ts files and any addition to them results in more maintenance work for me.

For the sake of continuity of the project, I need to be careful with PRs that create more maintenance work. Even if I have no experience with TypeScript, it was clear to me that the first version of this PR was going to create more work. 

The current version feels much better but it still worried me and that's why I looked for more opinions (thanks @DavidPeicho!)

So it sounds like I should not have merged the `Material | Material[]` change?

**Comment by mrdoob** [comment]
> It would be silly to revert `Material | Material[]` **only** to stay coherent with your decision regarding my PR.

It would be to stay coherent with this PR. It would be to avoid more PRs like this one.

**Comment by squidfunk** [comment]
If I may add my 2 cents – I'm perfectly fine using the "hack" as described in my Tweet (linked in OP). I agree that it may be difficult to map the inherently dynamic nature of Three.js into generics. As I understand it, `Object3D` is essentially a tree data structure with arbitrary types as child nodes, and my experience with TypeScript tells me that it's impossible to map this into a generic parameter. Thus, if this PR is rejected, I agree that removing the typings @mrdoob mentioned would be a logical consequence.

**Comment by mrdoob** [comment]
Okay, for now I think it's best to close this PR.
Thanks though!

**Comment by xawill** [comment]
You're welcome; happy to contribute where/when I can. That being said, I would have appreciated to have this kind of discussion earlier in the PR submission process. But this is a detail.

**Comment by mrdoob** [comment]
I'm sorry I was not able to review this PR earlier. I'm just a human being trying to handle hundreds of contributions.

**Comment by mrdoob** [comment]
#21174
